src/main/java/io/jenkins/plugins/credentials/secretsmanager/factory/CredentialsFactory.java

```diff
@@ -0,0 +1,153 @@
+import io.jenkins.plugins.credentials.secretsmanager.factory.string.AwsStringCredentials;
+import io.jenkins.plugins.credentials.secretsmanager.factory.username_password.AwsUsernamePasswordCredentials;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Supplier;
+import java.util.logging.Logger;
+
+public abstract class CredentialsFactory {
+
+    private CredentialsFactory() {
+
+    }
+
+    /**
+     * Construct a Jenkins credential from a Secrets Manager secret.
+     *
+     * @param name the secret's name (must be unique)
+     * @param description the secret's description
+     * @param tags the secret's AWS tags
+     */
+    public static Optional<StandardCredentials> create(String name, String description, Map<String, String> tags, AWSSecretsManager client) {
+        final String type = tags.getOrDefault("jenkins:credentials:type", "");
+        final String username = tags.getOrDefault("jenkins:credentials:username", "");
+
+        final CredentialType t = CredentialType.fromString(type);
+
+        switch (t) {
+            case string:
+                return Optional.of(new AwsStringCredentials(name, description, new SecretSupplier(client, name)));
+            case usernamePassword:
+                return Optional.of(new AwsUsernamePasswordCredentials(name, description, new SecretSupplier(client, name), username));
+            case sshUserPrivateKey:
+                return Optional.of(new AwsSshUserPrivateKey(name, description, new StringSupplier(client, name), username));
+            case certificate:
+                return Optional.of(new AwsCertificateCredentials(name, description, new SecretBytesSupplier(client, name)));
+            default:
+                return Optional.empty();
+        }
+    }
+
+    private static class SecretBytesSupplier extends RealSecretsManager implements Supplier<SecretBytes> {
+
+        private SecretBytesSupplier(AWSSecretsManager client, String name) {
+            super(client, name);
+        }
+
+        @Override
+        public SecretBytes get() {
+            return getSecretValue().match(new SecretValue.Matcher<SecretBytes>() {
+                @Override
+                public SecretBytes string(String str) {
+                    return null;
+                }
+
+                @Override
+                public SecretBytes binary(byte[] bytes) {
+                    return SecretBytes.fromBytes(bytes);
+                }
+            });
+        }
+    }
+
+    private static class SecretSupplier extends RealSecretsManager implements Supplier<Secret> {
+
+        private SecretSupplier(AWSSecretsManager client, String name) {
+            super(client, name);
+        }
+
+        @Override
+        public Secret get() {
+            return getSecretValue().match(new SecretValue.Matcher<Secret>() {
+                @Override
+                public Secret string(String str) {
+                    return Secret.fromString(str);
```

Comment:
This was what caused my null pointer problem.  Basically the call to Secret.fromString was being delayed until the agent called getPassword on the credential snapshot and then it blows up.

Reply:
Yep, that one was hard to find. Did you mean that my changes will fix it, or that they currently still cause an NPE?